src/tacacs/nss/patch/0010-Send-remote-address-in-TACACS-authorization-message.patch

```diff
@@ -0,0 +1,88 @@
+From aec0ad4da85b0edbfa45f6deab1c0abbeecd4bbb Mon Sep 17 00:00:00 2001
+From: liuh-80 <liuh@microsoft.com>
+Date: Wed, 28 Sep 2022 17:12:18 +0800
+Subject: [PATCH] Send remote address in TACACS+ authorization message.
+
+---
+ nss_tacplus.c | 53 ++++++++++++++++++++++++++++++++++++++++++++++++++-
+ 1 file changed, 52 insertions(+), 1 deletion(-)
+
+diff --git a/nss_tacplus.c b/nss_tacplus.c
+index 2de00a6..00b0d17 100644
+--- a/nss_tacplus.c
++++ b/nss_tacplus.c
+@@ -33,6 +33,7 @@
+ #include <ctype.h>
+ #include <netdb.h>
+ #include <nss.h>
++#include <limits.h>
+ 
+ #include <libtac/libtac.h>
+ 
+@@ -717,6 +718,44 @@ connect_tacacs(struct tac_attrib **attr, int srvr)
+     return fd;
+ }
+ 
++/*
++ * Get current SSH session remote address from environment variable
++ */
++void get_remote_address(char* dst, socklen_t size)
++{
++    memset(dst, 0, size);
++    char format[32];
++    snprintf(format, sizeof(format), "%%%ds", (int)(size-1));
++
++    // SSHD will create environment variable SSH_CONNECTION after user session created.
++    const char* ssh_connection = getenv("SSH_CONNECTION");
++    if (ssh_connection != NULL) {
++        // The first part of $SSH_CONNECTION is client IP address
++        sscanf(ssh_connection , format, dst);
++        if(debug) {
++            syslog(LOG_DEBUG, "%s: remote address=%s, SSH_CONNECTION=%s", nssname, ssh_connection, dst);
++        }
++        return;
++    }
++
++    if(debug) {
++        syslog(LOG_DEBUG, "%s: can't get environment variable SSH_CONNECTION, errno=%d", nssname, errno);
++    }
++
++    // Before user session created, SSHD will create environment variable SSH_REMOTE_IP.
++    const char* ssh_remote_ip = getenv("SSH_REMOTE_IP");
++    if (ssh_remote_ip != NULL) {
++        snprintf (dst , size, "%s", ssh_remote_ip);
```

Comment:
>snprintf [](http://example.com/codeflow?start=9&length=8)

I think you use `snprintf` to prevent buffer overrun. If that is the case, better to return nothing and let caller know there is something wrong, instead of returning a partial IpAddr string.

If we agree on, just compare strlen before strcpy.

Reply:
Fixed, check environment variable size before copy.